### PR TITLE
REF: reindex_indexer use np.void to avoid JoinUnit

### PR DESCRIPTION
**Pull request by jbrockmendel**
ATM we do a kludgy thing where we sometimes reindex before calling concat (e.g. in DataFrame.append) and that leads to all-NA columns that we choose to ignore when choosing the dtype for the concatenated column.  This leads to a) funky behavior when we encounter an "real" all-NA column and b) a big performance hit in checking for all-NA columns.

Instead of doing that, this PR implements the keyword use_na_proxy (chosen to match ArrayManager's keyword) that ends up creating ndarrays with np.void dtype instead of all-NA float64 ndarrays.

This will allow us to simplify a bunch of code in internals.concat (e.g. get rid of case where `unit.block is None`) and improve the funky behavior mentioned above.  Both of these are left to follow-ups to keep the diff manageable.

### CODE REVIEW

File: pandas/core/internals/concat.py [hunk 1]
```diff
@@ -198,6 +198,27 @@ def concatenate_managers(
     if isinstance(mgrs_indexers[0][0], ArrayManager):
         return _concatenate_array_managers(mgrs_indexers, axes, concat_axis, copy)
 
+    new_mgrs_indexers = []
```
Comment (jreback) [inline pandas/core/internals/concat.py:201]: i'd feel better if you make this a separate function and just call it here

Comment (jbrockmendel) [reply pandas/core/internals/concat.py:201]: so just refactor out L201-L220 as a new function?  sure

File: pandas/core/internals/concat.py [hunk 2]
```diff
@@ -245,6 +247,38 @@ def concatenate_managers(
     return BlockManager(tuple(blocks), axes)
 
 
+def _reindex_columns_void(
```
Comment (jreback) [inline pandas/core/internals/concat.py:250]: i would rename to `_maybe_reindex_columns_na_proxy`, e.g. don't use the term `void` as we don't use it elsewhere and instead use `na_proxy` (I know the dtype is V which is fine); alt make this clear in the doc-string here.

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
greenish, just test_to_csv_compression_encoding_gcs

File: pandas/core/internals/concat.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/internals/concat.py]: changed + green

**Pull request merged**